prost-build/src/code_generator.rs

```diff
@@ -1036,6 +1016,38 @@ impl<'a> CodeGenerator<'a> {
         }
     }
 
+    /// Returns whether the Rust type for this field needs to be `Box<_>`.
+    ///
+    /// This can be explicitly configured with `Config::boxed`, or necessary
+    /// to prevent an infinitely sized type definition in case when the type of
+    /// a non-repeated message field transitively contains the message itself.
+    fn boxed(
+        &self,
+        field: &FieldDescriptorProto,
+        fq_message_name: &str,
+        oneof: Option<&str>,
+    ) -> bool {
+        if field.label == Some(Label::Repeated as i32) {
+            return false;
+        }
+        let fd_type = field.r#type();
+        if (fd_type == Type::Message || fd_type == Type::Group)
+            && self
+                .message_graph
+                .is_nested(field.type_name(), fq_message_name)
+        {
+            return true;
+        }
+        let config_path = match oneof {
+            None => Cow::Borrowed(fq_message_name),
+            Some(ooname) => Cow::Owned(format!("{fq_message_name}.{ooname}")),
+        };
```

Comment:
I have a few points that are related:
- I don't like the concat of these names here. I understand the original code uses fq_message_name (few lines up) and the oneof_name (which is now this concat).
- The primary function of MessageGraph is to be used in this boxed check. However, MessageGraph is not aware of the boxed definition. Maybe it makes sense to move this boxed definition inside MessageGraph?
- MessageGraph can mark messages as recursive, but that doesn't consider manually boxed fields.
- Why are oneof fields not added to MessageGraph? If that was true, is_nested can also work with oneof_name and this function only needs a single name.

I think this boxed and MessageGraph code could use a refactor. I don't have concrete advice to give. Are you willing to dive into this?

Reply:
I could improve MessageGraph in a later PR, this one only moves code around without changing the functionality (apart from the logic change on repeated fields which I will revert).

Reply:
I agree that the `MessageGraph` should be changed in another PR.